### PR TITLE
Track deployment namespaces in Deployer

### PR DESCRIPTION
**Pull request by nkubala**
**Related**: https://github.com/GoogleContainerTools/skaffold/pull/6117

**Description**

Currently, the `Deploy()` method defined on all `Deployer` implementations returns two values: any `error` seen during the deployment, and a list of `namespaces` deployed to. This list of namespaces is then returned back to the `Runner`, which updates them on the `RunContext` to be used later. This pattern is:
1) Unnecessary, because all of the components that use `runContext.GetNamespaces()` live inside the `Deployer` implementation, and 
2) Problematic, because it defines a messy API for future `Deployer` implementations that have no use for the concept of a namespace

This change moves the tracking of all deployed namespaces into the `Deployer`, which can then feed those namespaces to any subcomponents that require them to function properly.

This changes the `Deploy` call to a very simple API:
```golang
Deploy(context.Context, io.Writer, []graph.Artifact) error
```

Additionally, all sub-components that provide a `Start()` method for orchestration by the `Runner` (`Accessor`, `Debugger`, `Monitor`, and `Logger` have had the `namespaces` parameter removed from their signatures. Using `Accessor` as an example:
```golang
type Accessor interface {
  Start(context.Context, io.Writer)
}
```

All Kubernetes-based `Deployer` implementations provide a `trackNamespaces()` method, called at the end of a `Deploy()`, which will update an internal `[]string` with the new namespaces. This `[]string` is then passed by reference to all subcomponents, which can consume whatever the contents are as the effective namespaces.

*NOTE: We can safely use a `*[]string` to represent the active namespaces, as we know that no subcomponents will be activated by the `Runner` until the `Deploy()` call exits successfully.*

**This change is a non-functional refactor.**

### CODE REVIEW

File: pkg/skaffold/deploy/deploy_mux.go [hunk 1]
```diff
@@ -99,33 +99,30 @@ func (m DeployerMux) RegisterLocalImages(images []graph.Artifact) {
...
-
+func (m DeployerMux) Deploy(ctx context.Context, w io.Writer, as []graph.Artifact) error {
 	for i, deployer := range m.deployers {
 		eventV2.DeployInProgress(i)
 		w = output.WithEventContext(w, constants.Deploy, strconv.Itoa(i), "skaffold")
 		ctx, endTrace := instrumentation.StartTrace(ctx, "Deploy")
 
-		namespaces, err := deployer.Deploy(ctx, w, as)
-		if err != nil {
+		var err error
```
Comment (MarlonGamez) [inline pkg/skaffold/deploy/deploy_mux.go:108]: nit: any reason to not just use `:=` on 109 and 115?

File: pkg/skaffold/deploy/util/util_test.go [hunk 2]
```diff
@@ -79,13 +79,9 @@ func TestRunContext_UpdateNamespaces(t *testing.T) {
 	}
 	for _, test := range tests {
 		testutil.Run(t, test.description, func(t *testutil.T) {
-			runCtx := &RunContext{
-				Namespaces: test.oldNamespaces,
-			}
+			ns := ConsolidateNamespaces(test.oldNamespaces, test.newNamespaces)
```
Comment (MarlonGamez) [inline pkg/skaffold/deploy/util/util_test.go:82]: nit: Should we rename the test function to reflect `ConsolidateNamespaces()`?

**Review by tejal29 — CHANGES_REQUESTED** [review]
Can we get rid of all the `Provider` interfaces before this. That would help clean up few interfaces.

**Review by MarlonGamez — COMMENTED** [review]
couple small nits, nothing major

File: pkg/skaffold/deploy/util/util_test.go (hunk 2 above)
Comment (nkubala) [reply pkg/skaffold/deploy/util/util_test.go:82]: copy paste error 😅  thanks

File: pkg/skaffold/deploy/deploy_mux.go (hunk 1 above)
Comment (nkubala) [reply pkg/skaffold/deploy/deploy_mux.go:108]: nope I should have done that, updated

**Comment by nkubala** [comment]
> Can we get rid of all the Provider interfaces before this. That would help clean up few interfaces.

@tejal29 should that block this change? not saying we shouldn't clean it up, but I'm not convinced it needs to happen before this, it seems a bit orthogonal.

**Comment by tejal29** [comment]
> All Kubernetes-based `Deployer` implementations provide a `trackNamespaces()` method, called at the end of a `Deploy()`, which will update an internal `[]string` with the new namespaces. This `[]string` is then passed by reference to all subcomponents, which can consume whatever the contents are as the effective namespaces.

Is this `namespace` now part of `Deploy` Object. Can we remove it from RunContext?
Also a bunch of methods in PortForwarder, Logger and Debugger have changed right?
Can you list them in the description

**Comment by tejal29** [comment]
> > Can we get rid of all the Provider interfaces before this. That would help clean up few interfaces.
> 
> @tejal29 should that block this change? not saying we shouldn't clean it up, but I'm not convinced it needs to happen before this, it seems a bit orthogonal.

The number of unnecessary changes in this PR will reduce. If we do them later, we will be adding a bunch of changes and then deleting them in the later PR.

**Comment by nkubala** [comment]
@tejal29

> Is this namespace now part of Deploy Object. Can we remove it from RunContext?
Good catch, meant to do this. Updated

> Also a bunch of methods in PortForwarder, Logger and Debugger have changed right? Can you list them in the description?
Done.

https://github.com/GoogleContainerTools/skaffold/pull/6190 has also been merged to remove the `Provider` interfaces, which should make this PR easier to review.

File: pkg/skaffold/deploy/deploy.go [hunk 3]
```diff
@@ -32,8 +32,8 @@ import (
 // the build results to a Kubernetes cluster
 type Deployer interface {
 	// Deploy should ensure that the build results are deployed to the Kubernetes
-	// cluster. Returns the list of impacted namespaces.
-	Deploy(context.Context, io.Writer, []graph.Artifact) ([]string, error)
+	// cluster.
+	Deploy(context.Context, io.Writer, []graph.Artifact) error
```
Comment (tejal29) [inline pkg/skaffold/deploy/deploy.go:36]: yay!

**Review by tejal29 — COMMENTED** [review]
LGMT. 

Testing deploy now.

**Comment by tejal29** [comment]
# Testing Notes
1) `trackNamespaces` is called in all 4 deployers. 
image

2) Debug works for cloudcode-samples/nodejs/ in default namespace.
 - [ ] debugging break point hit. 
 - [ ] portfowarded urls working.

3) Debug works for cloudcode-samples/nodejs/ in a `test` namespace using `SKAFFOLD_NAMESPACE` environment variable in debug config.
 - [ ] debugging break point hit. 
 - [ ] portfowarded urls working.

4) Testing helm deplyer
```
➜  helm-deployment git:(deploy-namespaces) ✗ skaffold dev --namespace test
 - test:deployment/skaffold-helm is ready.
Deployments stabilized in 2.235 seconds
Press Ctrl+C to exit
Watching for changes...
[skaffold-helm] Hello world! 0
[skaffold-helm] Hello world! 1
[skaffold-helm] Hello world! 0
[skaffold-helm] Hello world! 1
[skaffold-helm] Hello world! 2
[skaffold-helm] Hello world! 2
```

**Review by tejal29 — APPROVED** [review]

**Comment by gsquared94** [comment]
a bad merge seems to have reverted https://github.com/GoogleContainerTools/skaffold/pull/6135

**Pull request merged**